### ISSUE TITLE
Internal Error when compiling let

### ISSUE THREAD

**Issue by redfish64** [idris-lang/Idris2#1096]
https://gist.github.com/redfish64/08df655d0ddf8a4ff2915538f4d4f671

```
foo : Int
foo = let x : (Int, Int)
          x@(a,b) = (42,43)
      in a
```

```
$ idris2 LetInternalError.idr 
     ____    __     _         ___                                           
    /  _/___/ /____(_)____   |__ \                                          
    / // __  / ___/ / ___/   __/ /     Version 0.3.0-52203baaf
  _/ // /_/ / /  / (__  )   / __/      https://www.idris-lang.org           
 /___/\__,_/_/  /_/____/   /____/      Type :? for help                     

Welcome to Idris 2.  Enjoy yourself!
1/1: Building LetInternalError (LetInternalError.idr)
Error: INTERNAL
ERROR: x@((|((Builtin.Pair $a) $b),((Builtin.MkPair $a) $b)|)) is not a function application
Main> 

```

**Comment by ohad** [comment]
Nice find!

As far as I know, this program should not be syntactically valid (though I wish it was!): we do not bind multiple names in one local definition.

If this is blocking you, you can try using:
```idris
foo : Int
foo = let x@(a,b) : (Int, Int) = (42,43)
      in a
```
and now `x,a,b` can be used in the body, for example:
```idris
(+) : (Int, Int) -> (Int, Int) -> (Int, Int)
(+) (x1, y1) (x2, y2) = (x1 + x2, y1 + y2)

bar : (Int,Int)
bar = let x@(a,b) : (Int, Int) = (42,43)
      in (a+1, b+1) + x
```

There is one case where this workaround will not work: when the type-checker needs the values of `a,x,y` to type-check the body of the `let`, for example:

```idris
gnu : (u : (Int, Int) ** (v : Int ** (fst u = v, v = 42)))
gnu = let x@(a,b) : (Int, Int) = (42,43)
      in (x ** (a ** (Refl, Refl)))

 --  While processing right hand side of gnu. When unifying 42 and ?fst [locals in scope: a, b].
     Mismatch between: 42 and a.
     
     Internal.idr:3:29--3:33
      1 | gnu : (u : (Int, Int) ** (v : Int ** (fst u = v, v = 42)))
      2 | gnu = let x@(a,b) : (Int, Int) = (42,43)
      3 |       in (x ** (a ** (Refl, Refl)))
                                      ^^^^
```

In that case, you will need to use the much more verbose workaround:
```idris
gnu : (u : (Int, Int) ** (v : Int ** (fst u = v, v = 42)))
gnu = let x : (Int, Int)
          x = (42,43)
          a : Int
          a = fst x
          b : Int
          b = snd x
      in (x ** (a ** (Refl, Refl)))
```

**Comment by gallais** [comment]
This was fixed in #1816.
